### PR TITLE
Optional upstream dependencies

### PR DESCRIPTION
**Pull request by akshaydewan**
Issue #4931 - Optional upstream dependencies

Description:
Add a `skipScheduling` attribute to a dependency material. This can be used to prevent a Pipeline from triggering a downstream build.

### CODE REVIEW

**Comment by akshaydewan** [comment]
@maheshp This is still WIP as I'm adding some more test scenarios, but can you take a look and let me know if you have any feedback?

The main logic for skipping the scheduling is in `AutoBuild.java`. I had considered adding a `SchedulingChecker` implementation instead, but that wouldn't give me the material revisions that have changed.

For manual and timer triggers of the pipleine, the latest upstream dependency will be picked up.

File: server/src/main/java/com/thoughtworks/go/server/service/AutoBuild.java [hunk 1]
```diff
@@ -88,6 +89,10 @@ public void canProduce(PipelineConfig pipelineConfig, SchedulingCheckerService s
     public boolean isValidBuildCause(PipelineConfig pipelineConfig, BuildCause buildCause) {
         for (MaterialRevision materialRevision : buildCause.getMaterialRevisions()) {
             if (materialRevision.isChanged()) {
+                if (materialRevision.getMaterial() instanceof DependencyMaterial) {
+                    boolean skipSchedulingOnChange = ((DependencyMaterial) (materialRevision.getMaterial())).skipSchedulingOnChange();
+                    return !skipSchedulingOnChange;
```
Comment (arvindsv) [inline server/src/main/java/com/thoughtworks/go/server/service/AutoBuild.java]: This is just a drive-by comment. Suppose there are two dependency materials which are considered changed (a simple diamond dependency will do), and if the first one is set to skip scheduling on change and the second one is set to not skip scheduling on change, will this work?

Comment (akshaydewan) [reply server/src/main/java/com/thoughtworks/go/server/service/AutoBuild.java]: Yeah, Mahesh pointed that out yesterday. I'll be changing the logic and adding tests to verify it.

**Comment by akshaydewan** [comment]
In this PR, the only way to set `skipScheduling` is by editing the XML. We can make the API and UI changes for pipeline config separately

**Comment by arvindsv** [comment]
One thing to watch out for, that I've seen happen in cases where we allow editing by XML only is: You go and edit the XML. Everything looks fine. Then, you go and edit something else in the UI. Your XML change gets overwritten.

**Comment by akshaydewan** [comment]
> One thing to watch out for, that I've seen happen in cases where we allow editing by XML only is: You go and edit the XML. Everything looks fine. Then, you go and edit something else in the UI. Your XML change gets overwritten.

Thanks for pointing it out. Verified this, the xml isn't getting overwritten.

File: domain/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterial.java [hunk 2]
```diff
@@ -229,4 +229,9 @@ public void updateFromConfig(MaterialConfig materialConfig) {
     @Override
     public void populateAgentSideEnvironmentContext(EnvironmentVariableContext context, File workingDir) {
     }
+
+    @Override
+    public boolean skipSchedulingOnChange() {
```
Comment (maheshp) [inline domain/src/main/java/com/thoughtworks/go/config/materials/AbstractMaterial.java]: Probably add a test for this with the reason why its always false?

**Comment by maheshp** [comment]
> One thing to watch out for, that I've seen happen in cases where we allow editing by XML only is: You go and edit the XML. Everything looks fine. Then, you go and edit something else in the UI. Your XML change gets overwritten.

>>Thanks for pointing it out. Verified this, the xml isn't getting overwritten.

Editing through UI does not overwrite, but the changes will be lost if using the PipelineConfig API V9, as mentioned above PipelineConfig API needs to bumped up to support `skipScheduling`
